### PR TITLE
Lifecycle fixes

### PR DESCRIPTION
**Pull request by sjcorbett**
Fixes a handful of surprising behaviours in AbstractApplication, DynamicGroup and BasicStartable.

### CODE REVIEW

**Comment by sjcorbett** [comment]
One possibly surprising new behaviour from 1e44522 involves the status of a `BasicStartable` when its children restart. I'll use this blueprint as an example:
```yaml
location: localhost
services:
- type: org.apache.brooklyn.entity.stock.BasicStartableImpl
  brooklyn.children:
  - type: org.apache.brooklyn.entity.software.base.EmptySoftwareProcess
    brooklyn.config:
      postLaunchCommand: "exit 1"
      startTimeout: 2s
```
Previously, once the ESP entity has failed once the `BasicStartable` is _always_ on fire, even if `postLaunchCommand` were modified to something sensible. With the changes in this pull request when the ESP restarts the BasicStartable's state transitions to RUNNING, then back to `ON_FIRE` when the post-launch command fails for the second time. I discussed this with @justin.thompson. We felt that though the new behaviour isn't entirely satisfactory it is less wrong than before.

File: core/src/main/java/org/apache/brooklyn/entity/stock/BasicStartableImpl.java [hunk 1]
```diff
@@ -51,6 +53,7 @@
     public void start(Collection<? extends Location> locations) {
         try {
             ServiceStateLogic.setExpectedState(this, Lifecycle.STARTING);
+            ServiceProblemsLogic.clearProblemsIndicator(this, START);
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/entity/stock/BasicStartableImpl.java]: Any specific reason to add this? Don't see it being set elsewhere.

File: software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessDriver.java [hunk 2]
```diff
@@ -312,16 +312,36 @@ public void run() {
...
+                }
             }});
             DynamicTasks.queue("post-launch", new Runnable() { public void run() {
-                postLaunch();
+                try {
+                    postLaunch();
+                } catch (Exception e) {
+                    ServiceStateLogic.setExpectedState(getEntity(), Lifecycle.ON_FIRE);
+                    throw Exceptions.propagate(e);
+                }
```
Comment (neykov) [inline software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessDriver.java]: Still need to do the same for the `doFullStartOnRestart()` case. Suggest instead wrapping the `restart` effector as done in [`start`](https://github.com/apache/brooklyn-server/blob/master/software/base/src/main/java/org/apache/brooklyn/entity/software/base/lifecycle/MachineLifecycleEffectorTasks.java#L353-L363).

**Comment by sjcorbett** [comment]
@neykov I've updated per your comments.

**Comment by neykov** [comment]
Thanks @sjcorbett, merging.

**Pull request merged**